File: worker/rdb/consumerGroup.go
package rdb

import (
	"context"
	"log"
	"strings"

	"github.com/redis/go-redis/v9"
)

func CreateConsumerGroup(ctx context.Context, client *redis.Client, stream, consumerGroup string) {

	err := client.XGroupCreateMkStream(ctx, stream, consumerGroup, "0").Err()
	if err != nil {
		if strings.Contains(err.Error(), "BUSYGROUP") {
			log.Println("Consumer group already exists")
		} else {
			log.Fatalf("Error creating consumer group: %v", err)
		}
	} else {
		log.Println("Consumer group created")
	}
}

File: discordbot/src/rdb/consumer-group.ts
import { RedisClientType } from "redis";

/** Ensure a consumer-group exists (no-op if it already does) */
export async function createConsumerGroup(
  redis: RedisClientType,
  stream: string,
  consumerGroup: string
): Promise<void> {
  try {
    await redis.xGroupCreate(stream, consumerGroup, "0", { MKSTREAM: true });
    console.log(`[outStream] ${consumerGroup} created on ${stream}`);
  } catch (err: any) {
    if (!String(err?.message).includes("BUSYGROUP")) throw err;
    console.log(`[outStream] consumerGroup: ${consumerGroup} already exists`);
  }
}

/** Read *new* entries for this consumer, blocking up to `blockMs` */
export async function readGroup(
  redis: RedisClientType,
  stream: string,
  group: string,
  consumer: string,
  blockMs = 0,
  count = 10
) {
  return redis.xReadGroup(
    group,
    consumer,
    { key: stream, id: ">" },          // “>” means only new entries
    { BLOCK: blockMs, COUNT: count }
  );
}

export function ack(
  redis: RedisClientType,
  stream: string,
  group: string,
  ids: string[]
) {
  return redis.xAck(stream, group, ids);
}
